app/models/prompt.rb

```diff
@@ -120,18 +120,23 @@ def allowed_tags
       TagSet::TAG_TYPES.each do |tag_type|
         # if we have a specified set of tags of this type, make sure that all the
         # tags in the prompt are in the set.
-        if TagSet::TAG_TYPES_RESTRICTED_TO_FANDOM.include?(tag_type) && restriction.send("#{tag_type}_restrict_to_fandom")
-          # skip the check, these will be tested in restricted_tags below
-        elsif restriction.has_tags?(tag_type)
-          disallowed_taglist = tag_set.send("#{tag_type}_taglist") - restriction.tags(tag_type)
+
+        next if TagSet::TAG_TYPES_RESTRICTED_TO_FANDOM.include?(tag_type) && restriction.send("#{tag_type}_restrict_to_fandom")
+        # skip the check, these will be tested in restricted_tags below
+
+        taglist = tag_set.send("#{tag_type}_taglist")
+        next if taglist.empty?
+
+        if restriction.has_tags?(tag_type)
+          disallowed_taglist = taglist - restriction.tags(tag_type)
           unless disallowed_taglist.empty?
             errors.add(:base, ts("^These %{tag_label} tags in your %{prompt_type} are not allowed in this challenge: %{taglist}",
               tag_label: tag_type_label_name(tag_type).downcase,
               prompt_type: self.class.name.downcase,
               taglist: disallowed_taglist.collect(&:name).join(ArchiveConfig.DELIMITER_FOR_OUTPUT)))
           end
         else
-          noncanonical_taglist = tag_set.send("#{tag_type}_taglist").reject {|t| t.canonical}
+          noncanonical_taglist = taglist.reject {|t| t.canonical}
           unless noncanonical_taglist.empty?
             errors.add(:base, ts("^These %{tag_label} tags in your %{prompt_type} are not canonical and cannot be used in this challenge: %{taglist}. To fix this, please ask your challenge moderator to set up a tag set for the challenge. New tags can be added to the tag set manually by the moderator or through open nominations.",
               tag_label: tag_type_label_name(tag_type).downcase,
```

Comment:
From how I understand it, if the challenge has no restriction at all, we don't check for canonicity, but if it has any sort of restriction, we make sure all tags are (at least) canonical? Is that intended behavior?

Reply:
My knowledge of the challenge code is extremely rusty, but [the Tag Options part of the Gift Exchange FAQ](https://archiveofourown.org/faq/tutorial-running-a-gift-exchange-on-ao3?language_id=en#setgiftxtagopt) might help explain the intended/current behavior. The tl;dr is tags shouldn't need to be canonical if they're in one of the challenge's tag sets.